### PR TITLE
[IMP] l10n_pt: refactor Portuguese localization

### PR DESCRIPTION
**Pull request by rigr-odoo**
This PR completely refactors and updates the Portuguese localization to include the:
- Chart of Accounts (the accounts as well as the account groups)
- Fiscal positions
- Taxes (including the ones from the islands of Madeira and Azores)
- Tax groups
- Tax report according to the Governement template
- Translations files

Links:
CoA: https://www.occ.pt/fotos/editor2/taxonomiasplanocontas_fev2019.pdf
EU reverse charge: https://europa.eu/youreurope/business/taxation/vat/cross-border-vat/index_en.htm
Tax report template: https://info.portaldasfinancas.gov.pt/pt/apoio_contribuinte/modelos_formularios/iva/Documents/declaracao_periodica_IVA.pdf
Information on how to complete the tax report: https://www.occ.pt/news/_manual/ColecaoIVA2019.pdf

### CODE REVIEW

File: addons/l10n_pt/data/account.group.template.csv [hunk 1]
```diff
@@ -0,0 +1,146 @@
+id,code_prefix_start,code_prefix_end,name,chart_template_id/id
```
Comment (h4818) [inline addons/l10n_pt/data/account.group.template.csv:1]: no lines have a `code_prefix_end`, do we need it?

**Review by h4818 — COMMENTED** [review]
Can't say I know anything about pt taxes

File: addons/l10n_pt/data/account.group.template.csv (hunk 1 above)
Comment (rigr-odoo) [reply addons/l10n_pt/data/account.group.template.csv:1]: Not needed as it is done here for instance https://github.com/odoo/odoo/blob/master/addons/l10n_be/data/account.group.template.csv :)

Comment (h4818) [reply addons/l10n_pt/data/account.group.template.csv:1]: But that example has a few `code_prefix_end`s. Does this file have any?

Comment (rigr-odoo) [reply addons/l10n_pt/data/account.group.template.csv:1]: Yes but only when the `code_prefix_end` is different from the `code_prefix_start` due to this:
https://github.com/odoo/odoo/blob/master/addons/account/models/account_account.py#L801-L804

Comment (h4818) [reply addons/l10n_pt/data/account.group.template.csv:1]: The `onchange` will not have any impact on an imported file. 
In cases where there are values in the field for at least one record, it makes sense to have it. But in this case, no records have a value in that field. 
There are many other l10n data files for account.group that have only the `_start` field, but not the `_end` - I am not sure why we need this field here if we're not providing any data.

File: addons/l10n_pt/data/account_chart_template_data.xml [hunk 2]
```diff
@@ -11,10 +11,12 @@
         <field name="property_account_income_categ_id" ref="chart_711"/>
         <field name="property_account_expense_categ_id" ref="chart_311"/>
         <field name="income_currency_exchange_account_id" ref="chart_7861"/>
-        <field name="expense_currency_exchange_account_id" ref="chart_692"/>
+        <field name="expense_currency_exchange_account_id" ref="chart_6928"/>
```
Comment (smetl) [inline addons/l10n_pt/data/account_chart_template_data.xml:14]: `chart_6928,6928,503,Others,expense,l10n_pt.pt_chart_template,False`
Others? It looks incorrect

File: addons/l10n_pt/data/account_chart_template_data.xml [hunk 3]
```diff
@@ -11,10 +11,12 @@
...
         <field name="default_pos_receivable_account_id" ref="chart_2117"/>
         <field name="account_journal_early_pay_discount_loss_account_id" ref="chart_682"/>
         <field name="account_journal_early_pay_discount_gain_account_id" ref="chart_728"/>
+        <field name="default_cash_difference_income_account_id" ref="chart_7861"/>
```
Comment (smetl) [inline addons/l10n_pt/data/account_chart_template_data.xml:18]: The same as exchange diff? I don't think it's correct

File: addons/l10n_pt/data/account_chart_template_data.xml [hunk 4]
```diff
@@ -11,10 +11,12 @@
...
+        <field name="default_cash_difference_expense_account_id" ref="chart_6863"/>
```
Comment (smetl) [inline addons/l10n_pt/data/account_chart_template_data.xml:19]: I think that one should be in `expense_currency_exchange_account_id`

File: addons/l10n_pt/data/account_tax_data.xml [hunk 5]
```diff
@@ -1,532 +1,1921 @@
 <?xml version="1.0" encoding="utf-8"?>
 <odoo>
     <data noupdate="1">
-        <record id="iva23" model="account.tax.template">
+
+<!--    Taxes Portugal Continental-->
```
Comment (smetl) [inline addons/l10n_pt/data/account_tax_data.xml:5]:
```suggestion
        <!-- Taxes Portugal Continental -->
```

Comment (smetl) [reply addons/l10n_pt/data/account_tax_data.xml:5]: Also, the `noupdate` should be removed I think

File: addons/l10n_pt/data/account_tax_data.xml [hunk 6]
```diff
@@ -1,532 +1,1921 @@
...
+
+                (0,0, {
+                    'factor_percent': 100,
+                    'repartition_type': 'tax',
+                    'account_id': ref('chart_2432'),
+                }),
+            ]"/>
+        </record>
+
+
```
Comment (smetl) [inline addons/l10n_pt/data/account_tax_data.xml:419]: plz remove the extra lines

File: addons/l10n_pt/data/account_tax_data.xml [hunk 7]
```diff
@@ -1,532 +1,1921 @@
...
+                    'repartition_type': 'tax',
+                    'account_id': ref('chart_2432'),
+                }),
+            ]"/>
+            <field name="refund_repartition_line_ids" eval="[(5, 0, 0),
+                (0,0, {
+                    'factor_percent': 100,
+                    'repartition_type': 'base',
+                }),
+
```
Comment (smetl) [inline addons/l10n_pt/data/account_tax_data.xml:410]: remove such lines between comments

File: addons/l10n_pt/data/account_tax_data.xml [hunk 8]
```diff
@@ -1,532 +1,1921 @@
...
+            <field name="description">VAT 0%</field>
+            <field name="sequence">107</field>
+            <field name="price_include" eval="0"/>
+            <field name="active" eval="False"/>
+            <field name="amount">0</field>
+            <field name="amount_type">percent</field>
+            <field name="type_tax_use">purchase</field>
+            <field name="tax_group_id" ref="tax_group_iva_0"/>
+            <field name="invoice_repartition_line_ids" eval="[(5, 0, 0),
+                (0,0, {
```
Comment (smetl) [inline addons/l10n_pt/data/account_tax_data.xml:394]:
```suggestion
                (0, 0, {
```
? :p

File: addons/l10n_pt/data/account_tax_data.xml [hunk 9]
```diff
@@ -1,532 +1,1921 @@
...
+            ]"/>
+        </record>
+
+
+        <record id="iva_pt_purchase_non_eu_isenta" model="account.tax.template">
+            <field name="chart_template_id" ref="pt_chart_template"/>
+            <field name="name">0% Non-EU</field>
+            <field name="description">VAT 0%</field>
+            <field name="sequence">108</field>
+            <field name="price_include" eval="0"/>
```
Comment (smetl) [inline addons/l10n_pt/data/account_tax_data.xml:425]: price_include is False by default
also, `eval="False"` is more correct

File: addons/l10n_pt/data/account_tax_data.xml [hunk 10]
```diff
@@ -1,532 +1,1921 @@
...
+            <field name="active" eval="False"/>
```
Comment (smetl) [inline addons/l10n_pt/data/account_tax_data.xml:426]: same here... active is false by default

File: addons/l10n_pt/data/account_tax_report.xml [hunk 11]
```diff
@@ -13,461 +13,537 @@
             </record>
         </field>
         <field name="line_ids">
-            <record id="tax_report_pt_base" model="account.report.line">
+            <record id="trp_base" model="account.report.line">
```
Comment (smetl) [inline addons/l10n_pt/data/account_tax_report.xml:16]: trp?... Tax Report... P...?

File: addons/l10n_pt/data/account_tax_report.xml [hunk 12]
```diff
@@ -13,461 +13,537 @@
...
-                    <record id="tax_report_pt_tax_total_pagar_formula" model="account.report.expression">
-                        <field name="label">balance</field>
-                        <field name="engine">aggregation</field>
-                        <field name="formula">tax_report_pt_tax_vendas_faturas.balance - tax_report_pt_tax_vendas_notas_credito.balance - tax_report_pt_tax_despesas_faturas.balance + tax_report_pt_tax_despesas_notas_credito.balance</field>
-                        <field name="subformula">if_above(EUR(0))</field>
-                    </record>
+            <record id="trp_imposto_favor_estado" model="account.report.line">
+                <field name="name">Tax in favor of the State</field>
+                <field name="children_ids">
+                        <record id="trp_imposto_favor_estado1" model="account.report.line">
```
Comment (smetl) [inline addons/l10n_pt/data/account_tax_report.xml:275]: indent

File: addons/l10n_pt/models/account_account.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
+from odoo import models, fields
+
+
+class AccountAccount(models.Model):
+    _inherit = "account.account"
+
+    l10n_pt_taxonomy_code = fields.Integer("Taxonomy code")
```
Comment (smetl) [inline addons/l10n_pt/models/account_account.py:7]:
```suggestion
    l10n_pt_taxonomy_code = fields.Integer(string="Taxonomy code")
```

File: addons/l10n_pt/models/account_account.py [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class AccountAccountTemplate(models.Model):
+    _inherit = "account.account.template"
+
+    l10n_pt_taxonomy_code = fields.Integer("Taxonomy code")
```
Comment (smetl) [inline addons/l10n_pt/models/account_account.py:13]:
```suggestion
    l10n_pt_taxonomy_code = fields.Integer(string="Taxonomy code")
```

File: addons/l10n_pt/models/chart_template.py [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+            return
+        self._update_cash_bank_accounts(company)
+
+    def _get_account_vals(self, company, account_template, code_acc, tax_template_ref):
+        vals = super()._get_account_vals(company, account_template, code_acc, tax_template_ref)
+        if account_template.l10n_pt_taxonomy_code:
+            vals['l10n_pt_taxonomy_code'] = account_template.l10n_pt_taxonomy_code
+        return vals
+
+    def _update_cash_bank_accounts(self, company):
```
Comment (smetl) [inline addons/l10n_pt/models/chart_template.py:19]: you don't need a separated method for that. Put the content directly in `_load`

File: addons/l10n_pt/models/chart_template.py [hunk 16]
```diff
@@ -0,0 +1,32 @@
+from odoo import models
+
+
+class AccountChartTemplate(models.Model):
+    _inherit = "account.chart.template"
+
+    def _load(self, company):
+        super()._load(company)
```
Comment (smetl) [inline addons/l10n_pt/models/chart_template.py:8]: this method is returning something

File: addons/l10n_pt/models/chart_template.py [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+            ('type', '=', 'cash'),
+        ], limit=1)
+        if cash_journal:
+            cash_journal.default_account_id.l10n_pt_taxonomy_code = 1
+        bank_journal = self.env['account.journal'].search([
+            ('company_id', '=', company.id),
+            ('type', '=', 'bank'),
+        ], limit=1)
+        if bank_journal:
+            bank_journal.default_account_id.l10n_pt_taxonomy_code = 2
```
Comment (smetl) [inline addons/l10n_pt/models/chart_template.py:32]: override `_create_bank_journals` instead

File: addons/l10n_pt/models/chart_template.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+        super()._load(company)
+        if company.account_fiscal_country_id.code != 'PT':
+            return
+        self._update_cash_bank_accounts(company)
+
+    def _get_account_vals(self, company, account_template, code_acc, tax_template_ref):
+        vals = super()._get_account_vals(company, account_template, code_acc, tax_template_ref)
+        if account_template.l10n_pt_taxonomy_code:
+            vals['l10n_pt_taxonomy_code'] = account_template.l10n_pt_taxonomy_code
+        return vals
```
Comment (smetl) [inline addons/l10n_pt/models/chart_template.py:17]:
```suggestion
        vals = super()._get_account_vals(company, account_template, code_acc, tax_template_ref)
        vals['l10n_pt_taxonomy_code'] = account_template.l10n_pt_taxonomy_code
        return vals
```

File: addons/l10n_pt/data/account_tax_report.xml (hunk 11 above)
Comment (rigr-odoo) [reply addons/l10n_pt/data/account_tax_report.xml:16]: Tax Report Portugal :)

File: addons/l10n_pt/models/chart_template.py [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+
+class AccountChartTemplate(models.Model):
+    _inherit = "account.chart.template"
+
+    def _get_account_vals(self, company, account_template, code_acc, tax_template_ref):
+        vals = super()._get_account_vals(company, account_template, code_acc, tax_template_ref)
+        if company.account_fiscal_country_id.code != 'PT':
+            return vals
+        vals['l10n_pt_taxonomy_code'] = account_template.l10n_pt_taxonomy_code
+        return vals
```
Comment (smetl) [inline addons/l10n_pt/models/chart_template.py:12]: `l10n_pt_taxonomy_code` is supposed to be set only on pt, right? So
```suggestion
    def _get_account_vals(self, company, account_template, code_acc, tax_template_ref):
        vals = super()._get_account_vals(company, account_template, code_acc, tax_template_ref)
        vals['l10n_pt_taxonomy_code'] = account_template.l10n_pt_taxonomy_code
        return vals
```

File: addons/l10n_pt/models/chart_template.py [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+            return super()._get_default_bank_journals_data()
+        return [{
+            'name': _('Cash'),
+            'type': 'cash',
+            'l10n_pt_taxonomy_code': 1,
+        }, {
+            'name': _('Bank'),
+            'type': 'bank',
+            'l10n_pt_taxonomy_code': 2,
+        }]
```
Comment (smetl) [inline addons/l10n_pt/models/chart_template.py:26]: why not use the super call?
```suggestion
    @api.model
    def _get_default_bank_journals_data(self):
        journal_values_list = super()._get_default_bank_journals_data()
        if self.env.company.account_fiscal_country_id.code == 'PT':
            for journal_values in journal_values_list:
                if journal_values['account_type'] == 'cash':
                    journal_values['l10n_pt_taxonomy_code'] = 1
                elif journal_values['account_type'] == 'bank':
                    journal_values['l10n_pt_taxonomy_code'] = 2
        return journal_values_list
```

File: addons/l10n_pt/models/chart_template.py [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+        bank_journals = self.env['account.journal']
+        for acc in self._get_default_bank_journals_data():
+            journal = self.env['account.journal'].create({
+                'name': acc['name'],
+                'type': acc['type'],
+                'company_id': company.id,
+            })
+            journal.default_account_id.l10n_pt_taxonomy_code = acc['l10n_pt_taxonomy_code']
+            bank_journals += journal
+        return bank_journals
```
Comment (smetl) [inline addons/l10n_pt/models/chart_template.py:40]: same here
```py
    def _create_bank_journals(self, company, acc_template_ref):
        journals = super()._create_bank_journals(company, acc_template_ref)
        if company.account_fiscal_country_id.code == 'PT':
            for journal in journals:
                if journal.type == 'cash':
                    journal.default_account_id.l10n_pt_taxonomy_code = 1
                elif journal.type == 'bank':
                    journal.default_account_id.l10n_pt_taxonomy_code = 2
        return journals
```

**Review by smetl — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**